README.md
# COVID-19 AUTO GENERATOR
The aim of this project was to create a pre-configured COVID-19 derogatory travel certificate (October 2020 version) generator to be used by one or more people.

Once configured, it allows you to generate an authorization in two clicks.

> Please note, you use this generator at your own risk, I decline all responsibility in the event of an error resulting in a penalty

## Demo 
![demo gif](https://kasty.github.io/demo-covid-auto-gen.gif "Demo covid19 auto generator")

Online demo on https://kasty.github.io/

## Usage
The application is already built. You can deploy the `public` folder on your webserver after having configured the user's data.
* Clone the repository
* Edit the config file at `public/users.json`
Example :
```
{
  "users": {
    "robert": {
      "profile": {
        "firstname": "Robert",
        "lastname": "Tudor",
        "birthday": "24/06/1978",
        "placeofbirth": "Nantes",
        "address": "35 rue de l\"épée",
        "zipcode": "44000",
        "city": "Nantes",
        "emoji": "🐼"
      },
      "settings": {
        "choices": [
          "famille",
          "achats",
          "sport_animaux",
          "sante",
          "handicap"
        ],
        "defaultChoice": "achats"
      }
    },
    "janine": {
      "profile": {
        "firstname": "Janine",
        "lastname": "Tudor",
        "birthday": "01/05/1975",
        "placeofbirth": "Biarritz",
        "address": "35 rue de l\"épée",
        "zipcode": "44000",
        "city": "Nantes",
        "emoji": "🐙"
      },
      "settings": {
        "choices": [
          "travail",
          "famille",
          "achats",
          "sport_animaux",
          "sante",
          "missions"
        ],
        "defaultChoice": "travail"
      }
    }
  },
  "defaultUser": "robert"
}
```
* Deploy the content of the `build` folder on your webserver
* It is advisable to add protection to the file so as not to expose your personal data. For example, you can add an .htaccess file

## Contribute
* Clone the repository
* Install the dependencies :
```
npm install
```
* Start the local node server :
```
npm start
```
See Create React App documentation below for more informations
* To build your app run
```
npm run build
```
Then host the `build` folder on a webserver and enjoy the app on your phone !


## About Create React App

This project was bootstrapped with [Create React App](https://github.com/facebook/create-react-app).

### Available Scripts

In the project directory, you can run:

#### `npm start`

Runs the app in the development mode.\
Open [http://localhost:3000](http://localhost:3000) to view it in the browser.

The page will reload if you make edits.\
You will also see any lint errors in the console.

#### `npm test`

Launches the test runner in the interactive watch mode.\
See the section about [running tests](https://facebook.github.io/create-react-app/docs/running-tests) for more information.

#### `npm run build`

Builds the app for production to the `build` folder.\
It correctly bundles React in production mode and optimizes the build for the best performance.

The build is minified and the filenames include the hashes.\
Your app is ready to be deployed!

See the section about [deployment](https://facebook.github.io/create-react-app/docs/deployment) for more information.

#### `npm run eject`

**Note: this is a one-way operation. Once you `eject`, you can’t go back!**

If you aren’t satisfied with the build tool and configuration choices, you can `eject` at any time. This command will remove the single build dependency from your project.

Instead, it will copy all the configuration files and the transitive dependencies (webpack, Babel, ESLint, etc) right into your project so you have full control over them. All of the commands except `eject` will still work, but they will point to the copied scripts so you can tweak them. At this point you’re on your own.

You don’t have to ever use `eject`. The curated feature set is suitable for small and middle deployments, and you shouldn’t feel obligated to use this feature. However we understand that this tool wouldn’t be useful if you couldn’t customize it when you are ready for it.

### Learn More

You can learn more in the [Create React App documentation](https://facebook.github.io/create-react-app/docs/getting-started).

To learn React, check out the [React documentation](https://reactjs.org/).


CHANGELOG.md
# 1.1.1
## November 20. 2020
### Fix
* Load data file depending on current location (handle subfolder)

# 1.1.0
## November 13. 2020
### Fix
* Display the minutes with a 0 from 1 to 9 minutes

### Changes
* The application can be used as it is in the `build` folder. The user's data is loaded asynchronously.

# 1.0.0
First version, the application needs to be build for each usage

.eslintrc.js
module.exports = {
  extends: ['react-app'],
  parser: '@typescript-eslint/parser',
  overrides: [
    {
      files: ['**/*.ts?(x)'],
      rules: {
        'import/no-anonymous-default-export': [1, {
          allowArray: false,
          allowArrowFunction: false,
          allowAnonymousClass: false,
          allowAnonymousFunction: false,
          allowCallExpression: true, // The true value here is for backward compatibility
          allowLiteral: false,
          allowObject: true,
        }],
        'comma-dangle': [1, 'always-multiline'],
        'func-call-spacing': [1, 'never'],
        'space-infix-ops': 1,
        quotes: [1, 'single', 'avoid-escape'],
        'no-trailing-spaces': [1, { 'ignoreComments': true }],
        'prefer-template': 1,
        'eol-last': [1, 'always'],
        'react/jsx-max-props-per-line': [1, { 'maximum': 2, 'when': 'always' }],
        'no-confusing-arrow': [2, { allowParens: false }],
        'key-spacing': [1, {
          beforeColon: false,
          afterColon: true,
        }],
        'no-unused-vars': 'off',
        '@typescript-eslint/no-unused-vars': 1,
        'object-curly-newline': [1, {
          ObjectExpression: { multiline: true, minProperties: 1 },
          ObjectPattern: { multiline: true, minProperties: 3 },
          ImportDeclaration: { multiline: true, minProperties: 3 },
          ExportDeclaration: { multiline: true, minProperties: 3 },
        }],
        'indent': [1, 2],
        'object-property-newline': [1, {
          allowAllPropertiesOnSameLine: false,
        }],
        'object-curly-spacing': [1, 'always'],
      },
    },
  ],
};

src/config.ts
import { ConfigType } from './config/types';

const config: ConfigType = {
  reasonsList: {
    travail: {
      title: 'au boulot',
      emoji: '🖥',
    },
    famille: {
      title: "m'occuper de la famille",
      emoji: '💪🏼',
    },
    achats: {
      title: 'faire les courses',
      emoji: '🛒',
    },
    sport_animaux: {
      title: "prendre l'air",
      emoji: ' ☔️',
    },
    sante: {
      title: 'chez le doc (glups)',
      emoji: '🤒',
    },
    handicap: {
      title: 'me déplacer en situation de handicap',
      emoji: '🦽',
    },
    convocation: {
      title: 'à une convocation',
      emoji: '📄',
    },
    missions: {
      title: "participer à une mission d'intérêt général",
      emoji: '🔧',
    },
    enfants: {
      title: "chercher les enfants à l'école",
      emoji: '🧸',
    },
  },
};

export default config;


src/config/types.d.ts
export type ConfigType = {
  reasonsList: ReasonListType;
}
export type UserConfigType = {
  defaultUser?: keyof UsersType;
  users: UsersType;
}
type PartialRecord<K extends keyof any, T> = {
  [P in K]?: T;
};

type ReasonFieldsType = {
  title: string;
  emoji?: string;
}
type ReasonsNameType = 'travail' | 'famille' | 'achats' | 'sport_animaux' | 'sante' | 'handicap' | 'convocation' | 'missions' | 'enfants';

type ReasonListType = PartialRecord<ReasonsNameType, ReasonFieldsType>;

export type UsersType = Record<string, UserType>;

export type UserType = {
  profile: {
    firstname: string;
    lastname: string;
    birthday: string; // format: JJ/MM/AAAA
    placeofbirth: string;
    address: string;
    zipcode: string;
    city: string;
    emoji?: string;
    datesortie?: string;
    heuresortie?: string;
  },
  settings: {
    choices: string[];
    defaultChoice: string;
  }
};


src/config/config.d.ts
export type configType = {
  reasonsList: ReasonListType;
  defaultUser?: keyof UsersType;
  users: UsersType;
}
type PartialRecord<K extends keyof any, T> = {
  [P in K]?: T;
};

type ReasonFieldsType = {
  title: string;
  emoji?: string;
}
type ReasonsNameType = "travail" | "famille" | "achats" | "sport_animaux" | "sante" | "handicap" | "convocation" | "missions" | "enfants";

type ReasonListType = PartialRecord<ReasonsNameType, ReasonFieldsType>;

type UsersType = Record<string, UserType>;

type UserType = {
  profile: {
    firstname: string;
    lastname: string;
    birthday: string; // format: JJ/MM/AAAA
    placeofbirth: string;
    address: string;
    zipcode: string;
    city: string;
    emoji?: string;
  },
  settings: {
    choices: string[];
    defaultChoice: string;
  }
}


src/config.sample.ts
import { configType } from './config/types';

const config: configType = {
  // all possible reasons
  reasonsList: {
    travail: {
      title: 'au boulot',
      emoji: '🖥',
    },
    famille: {
      title: "m'occuper de la famille",
      emoji: '💪🏼',
    },
    achats: {
      title: 'faire les courses',
      emoji: '🛒',
    },
    sport_animaux: {
      title: "prendre l'air",
      emoji: ' ☔️',
    },
    sante: {
      title: 'chez le doc (glups)',
      emoji: '🤒',
    },
    handicap: {
      title: 'me déplacer en situation de handicap',
      emoji: '🦽',
    },
    convocation: {
      title: 'à une convocation',
      emoji: '📄',
    },
    missions: {
      title: "participer à une mission d'intérêt général",
      emoji: '🔧',
    },
    enfants: {
      title: "chercher les enfants à l'école",
      emoji: '🧸',
    },
  },
  // you can add as many users as you want
  users: {
    robert: {
      profile: {
        firstname: 'Robert',
        lastname: 'Tudor',
        birthday: '24/06/1978',
        placeofbirth: 'Nantes',
        address: '35 rue de l\'épée',
        zipcode: '44000',
        city: 'Nantes',
        // only displayed in the select box
        emoji: '🐼',
      },
      settings: {
        // limit the list of choices for this user
        choices: [
          'famille',
          'achats',
          'sport_animaux',
          'sante',
          'handicap',
        ],
        // user's default choice selected (optional)
        defaultChoice: 'sport_animaux',
      },
    },
    janine: {
      profile: {
        firstname: 'Janine',
        lastname: 'Tudor',
        birthday: '01/05/1975',
        placeofbirth: 'Biarritz',
        address: '35 rue de l\'épée',
        zipcode: '44000',
        city: 'Nantes',
        emoji: '🐙',
      },
      settings: {
        choices: [
          'travail',
          'famille',
          'achats',
          'sport_animaux',
          'sante',
          'missions',
        ],
        defaultChoice: 'travail',
      },
    },
  },
  defaultUser: 'janine',
};

export default config;
